### ISSUE TITLE
podman play kube tires to fetch amd64 infra container on arm

### ISSUE THREAD

**Issue by agners** [containers/libpod#4877]
/kind bug

**Description**

Using `podman play kube` on a ARMv7 system seems to pick the wrong infra container.

**Steps to reproduce the issue:**

1. Build podman for ARM (e.g. using OpenEmbedded and the meta-virtualization layer)

2. Run `podman play kube your.yaml`.

**Describe the results you received:**

```
# podman play kube kube.yaml 
ERRO[0002] Error freeing pod lock after failed creation: no such file or directory 
Error: error adding Infra Container: unable to pull k8s.gcr.io/pause:3.1: unable to pull image: Image architecture mismatch: image uses "amd64", expecting "arm"
```

**Describe the results you expected:**

No error but starting the pod.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
# podman --version
podman version 1.7.0
```

**Output of `podman info --debug`:**

```
n# podman info --debug
debug:
  compiler: gc
  git commit: b7ce1157b00af09f4a09e39b377aa3abff46ee05
  go version: go1.13.5
  podman version: 1.7.0
host:
  BuildahVersion: 1.12.0
  CgroupVersion: v1
  Conmon:
    package: Unknown
    path: /usr/bin/conmon
    version: 'conmon version 2.0.1-dev, commit: 6ef63dfcc65d5401bc7cc7170d569cefcdadcb38'
  Distribution:
    distribution: '"torizon"'
    version: 0.0.0-devel-20200109121308-build.0
  MemFree: 789422080
  MemTotal: 1048190976
  OCIRuntime:
    name: runc
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc8
      commit: 425e105d5a03fabd737a126ad93d62a9eeede87f-dirty
      spec: 1.0.1-dev
  SwapFree: 0
  SwapTotal: 0
  arch: arm
  cpus: 2
  eventlogger: journald
  hostname: colibri-imx7-02965216
  kernel: 5.4.2-0.0.0-devel+git.a1a7b308314d
  os: linux
  rootless: false
  uptime: 20m 46.73s
registries:
  search:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ConfigFile: /etc/containers/storage.conf
  ContainerStore:
    number: 1
  GraphDriverName: overlay
  GraphOptions:
    overlay.mountopt: nodev
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 1
  RunRoot: /var/run/containers/storage
  VolumePath: /var/lib/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

OE built distro without package manager.

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by vrothberg** [comment]
Thanks for opening the issue, @agners! Could you rerun with `podman --log-level=debug` and share the logs?

**Comment by agners** [comment]
@vrothberg here you go:

```
# podman --log-level=debug play kube kube.yaml
DEBU[0000] Using conmon: "/usr/bin/conmon"
DEBU[0000] Initializing boltdb state at /var/lib/containers/storage/libpod/bolt_state.db
DEBU[0000] Using graph driver overlay
DEBU[0000] Using graph root /var/lib/containers/storage
DEBU[0000] Using run root /var/run/containers/storage
DEBU[0000] Using static dir /var/lib/containers/storage/libpod
DEBU[0000] Using tmp dir /var/run/libpod
DEBU[0000] Using volume path /var/lib/containers/storage/volumes
DEBU[0000] Set libpod namespace to ""
DEBU[0000] [graphdriver] trying provided driver "overlay"
DEBU[0000] overlay test mount with multiple lowers succeeded
DEBU[0000] overlay test mount indicated that metacopy is not being used
DEBU[0000] backingFs=extfs, projectQuotaSupported=false, useNativeDiff=true, usingMetacopy=false
DEBU[0000] Initializing event backend journald
DEBU[0000] using runtime "/usr/bin/runc"
WARN[0000] Error initializing configured OCI runtime crun: no valid executable found for OCI runtime crun: invalid argument
INFO[0000] Found CNI network podman (type=bridge) at /etc/cni/net.d/87-podman-bridge.conflist
WARN[0000] Default CNI network name podman is unchangeable
DEBU[0000] Initialized SHM lock manager at path /libpod_lock
DEBU[0000] Podman detected system restart - performing state refresh
INFO[0001] About to del CNI network lo (type=loopback)
INFO[0001] Got pod network &{Name:weston-container Namespace:weston-container ID:53864faa0c01b0362f68f61ba6ce5f98112073c38bcc33faa27a4962a18423a6 NetNS: Networks:[] RuntimeConfig:map[podman:{IP: MAC: PortMappings:[] Bandwidth:<nil> IpRanges:[]}]}
ERRO[0001] error loading cached network config: network "podman" not found in CNI cache
INFO[0001] About to del CNI network podman (type=bridge)
ERRO[0001] Error deleting network: missing prevResult from earlier plugin
ERRO[0001] Error while removing pod from CNI network "podman": missing prevResult from earlier plugin
ERRO[0001] Error refreshing container 53864faa0c01b0362f68f61ba6ce5f98112073c38bcc33faa27a4962a18423a6: missing prevResult from earlier plugin
DEBU[0001] Created cgroup path machine.slice/machine-libpod_pod_60234055cac5136fe68829df331bbd276ce137c554cd4a2642df17fc2107e683.slice for parent machine.slice and name libpod_pod_60234055cac5136fe68829df331bbd276ce137c554cd4a2642df17fc2107e683
DEBU[0001] Created cgroup machine.slice/machine-libpod_pod_60234055cac5136fe68829df331bbd276ce137c554cd4a2642df17fc2107e683.slice
DEBU[0001] Got pod cgroup as machine.slice/machine-libpod_pod_60234055cac5136fe68829df331bbd276ce137c554cd4a2642df17fc2107e683.slice
DEBU[0001] parsed reference into "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.mountopt=nodev]k8s.gcr.io/pause:3.1"
DEBU[0001] reference "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.mountopt=nodev]k8s.gcr.io/pause:3.1" does not resolve to an image ID
DEBU[0001] parsed reference into "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.mountopt=nodev]k8s.gcr.io/pause:3.1"
DEBU[0001] reference rewritten from 'k8s.gcr.io/pause:3.1' to 'k8s.gcr.io/pause:3.1'
DEBU[0001] Trying to pull "k8s.gcr.io/pause:3.1"
DEBU[0001] Credentials not found
DEBU[0001] Using registries.d directory /etc/containers/registries.d for sigstore configuration
DEBU[0001]  No signature storage configuration found for k8s.gcr.io/pause:3.1
DEBU[0001] Looking for TLS certificates and private keys in /etc/docker/certs.d/k8s.gcr.io
DEBU[0001] GET https://k8s.gcr.io/v2/
DEBU[0001] Ping https://k8s.gcr.io/v2/ status 401
DEBU[0001] GET https://k8s.gcr.io/v2/token?scope=repository%3Apause%3Apull&service=k8s.gcr.io
DEBU[0002] GET https://k8s.gcr.io/v2/pause/manifests/3.1
DEBU[0002] Using blob info cache at /var/lib/containers/cache/blob-info-cache-v1.boltdb
DEBU[0002] Source is a manifest list; copying (only) instance sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53 for current system
DEBU[0002] GET https://k8s.gcr.io/v2/pause/manifests/sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53
DEBU[0002] IsRunningImageAllowed for image docker:k8s.gcr.io/pause:3.1
DEBU[0002]  Using default policy section
DEBU[0002]  Requirement 0: allowed
DEBU[0002] Overall: allowed
DEBU[0002] Downloading /v2/pause/blobs/sha256:e11a8cbeda8688fdc3a68dc246d6c945aa272b29f8dd94d0ea370d30c2674042
DEBU[0002] GET https://k8s.gcr.io/v2/pause/blobs/sha256:e11a8cbeda8688fdc3a68dc246d6c945aa272b29f8dd94d0ea370d30c2674042
DEBU[0003] Error pulling image ref //k8s.gcr.io/pause:3.1: Image architecture mismatch: image uses "amd64", expecting "arm"
DEBU[0003] Removing pod cgroup machine.slice/machine-libpod_pod_60234055cac5136fe68829df331bbd276ce137c554cd4a2642df17fc2107e683.slice
ERRO[0003] Error freeing pod lock after failed creation: no such file or directory
ERRO[0003] error adding Infra Container: unable to pull k8s.gcr.io/pause:3.1: unable to pull image: Image architecture mismatch: image uses "amd64", expecting "arm"
```

**Comment by vrothberg** [comment]
That's an interesting one.

Let's first look at the manifest list:
`$ skopeo inspect --raw  docker://k8s.gcr.io/pause:3.1 | jq .`
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 527,
         "digest": "sha256:59eec8837a4d942cc19a52b8c09ea75121acc38114a2c68b98983ce9356b8610",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 527,
         "digest": "sha256:c84b0a3a07b628bc4d62e5047d0f8dff80f7c00979e1e28a821a033ecda8fe53",
         "platform": {
            "architecture": "arm",
            "os": "linux"
         }
      },
...
}
```

The pause image is provided for different architectures. Let's try to force-pull the arm image:
```
$ podman pull --override-arch=arm  k8s.gcr.io/pause:3.1
Trying to pull k8s.gcr.io/pause:3.1...
  Image architecture mismatch: image uses "amd64", expecting "arm"
Error: error pulling image "k8s.gcr.io/pause:3.1": unable to pull k8s.gcr.io/pause:3.1: unable to pull image: Image architecture mismatch: image uses "amd64", expecting "arm"
```

We're seeing the same issue as reported. Although we explicitly tried to pull the arm image, the config/manifest claims to be for amd64. Same when pulling via digest.

Note that we don't have any issues when pulling from Docker Hub. We can do the same for docker.io/library/busybox:latest where everything's working as expected.

I experimented a bit more and found that when puling from `k8s.gcr.io` we are getting different data when trying to pull the amd64 and arm image _but_ the arm config claims to be for amd64.

To make sure that's not a bug in our stack, I tried the same with `docker v18.09.8` which yields the very same results. After pulling, I inspected the image which claims to be for amd64.

`k8s.gcr.io/pause:latest` seems to only exist for `amd64` and no other architectures anymore.

I don't yet know what's exactly going on with the pause images but it's a bit disturbing.

@mtrmac @mrunalp @rhatdan @baude FYI

**Comment by vrothberg** [comment]
I'll open an issue over at Kubernetes and ask to update the pause image.

**Comment by dawud** [comment]
I'm affected by this on Fedora IoT 31.

Running `rpm-ostree rollback`, which, among other changes, rolls back `podman`, solves the issue for me.

```bash
  podman 2:1.7.0-2.fc31 -> 2:1.6.2-2.fc31
```

```bash
[root@rpi3bv12n001 ~]# rpm-ostree status
State: idle
AutomaticUpdates: stage; rpm-ostreed-automatic.timer: no runs since boot
Deployments:
● ostree://fedora-iot:fedora/stable/aarch64/iot
                   Version: 31.20200108.0 (2020-01-08T14:46:07Z)
                BaseCommit: 93402e9d43a7b7be21b65ca17b8b4bf3df3a964140d3e2e76d40051fe4da4dc6
              GPGSignature: Valid signature by 7D22D5867F2A4236474BF7B850CB390B3C3359C4
           LayeredPackages: zabbix-agent

  ostree://fedora-iot:fedora/stable/aarch64/iot
                   Version: 31.20200117.0 (2020-01-17T05:40:50Z)
                BaseCommit: 8953e1fe42520c93cb3c494e8b58e083170686bcca27973e93b2aab8fb2ecb52
              GPGSignature: Valid signature by 7D22D5867F2A4236474BF7B850CB390B3C3359C4
           LayeredPackages: zabbix-agent
```
 So it seems to be a bug introduced in `libpod`/`podman` somewhere between 1.6.2 and 1.7.0, and not a problem in the images.

**Comment by vrothberg** [comment]
@dawud, we are performing some more rigorous checks on images now. The changes have been integrated with Podman v1.7.0. Hence, it works with pre-1.7.0 versions of Podman. Although the images work with older versions, those images are still wrong. The config of the arm image claims to be for amd64 which is ultimately causing Podman v1.7.0 to error out.

**Comment by dawud** [comment]
Yes you are indeed right, that seems to be case.

`# skopeo inspect docker://docker.io/pihole/pihole:4.3.2-1_aarch64`

reports:

```bash
    "Architecture": "amd64",
```

**Comment by mheon** [comment]
I believe the consensus is that we are going to remove these checks, as they are not necessarily helpful. I don't think this fix will make it into 1.7.1 though.

**Comment by mtrmac** [comment]
https://github.com/containers/image/issues/799 for tracking.
